### ISSUE TITLE
OM allows new newsletter subscriptions even though the visitor has been confirmed as a subscriber.

### ISSUE THREAD

**Issue by ADDISON74** [OpenMage/magento-lts#2196]
OM allows new newsletter subscriptions even though the user has been confirmed as a subscriber. The issue relates to the subscription form, I did not check what if the address is subscribed already and a new account is created in which the subscription box is checked by creator.

**Step 1**: In OM-20.14 Backend I set the following values for Customer Newsletter
- Allow Guest Subscription = Yes
- Need to Confirm = Yes

![config](https://user-images.githubusercontent.com/8360474/172069323-13636bef-1f25-49ff-b91a-102bbe20b878.jpg)

**Step 2**: In OM-20.14 Frontend I filled in an email address in Newsletter input box. After pressing the [Subscribe] button page reloads with the following message "Confirmation request has been sent."

![confirmation_sent](https://user-images.githubusercontent.com/8360474/172069446-d6b2766a-92fc-441d-a338-6949cadec9cb.jpg)

**Step 3**: In guest Inbox I received the newsletter subscription confirmation email. 

![confirmation_received](https://user-images.githubusercontent.com/8360474/172069494-85882de9-3864-403e-8c21-ebda9f257267.jpg)

**Step 4**: After pressing [Confirm Subscription] the Frontend page loads with the following message "Your subscription has been confirmed". Here follows step 4.1 which is solved in PR #2195 which sends the confirmation of the successful subscription.

**Now comes the funny part.**

**Step 5**: I filled in again the same email address in Newsletter input box. After pressing the [Subscribe] button page reloads with the following message "Thank you for your subscription.". This is completely wrong because the email address is already subscribed.

![confirmation_thank_you](https://user-images.githubusercontent.com/8360474/172069532-431c3fcd-b0c5-41b8-8959-7a51da086299.jpg)

**Step 6**: In guest Inbox I received again the newsletter subscription confirmation email.

![confirmation_resent](https://user-images.githubusercontent.com/8360474/172069559-d6d728d3-937e-4d3e-9c0b-63bb3eaccaf0.jpg)

**Step 7**: After pressing [Confirm Subscription] the Frontend page loads with the following message "Your subscription has been confirmed".

![confirmation_confirmed_2](https://user-images.githubusercontent.com/8360474/172069569-9559d0e0-191a-4ef3-86f2-45a6b09c5e12.jpg)

If we go to step 5 we are already in an endless loop. This is a matter of logic, once an email address is subscribed OpenMage should display specific error messages, such as already subscribed. 

All the trouble is happening in this file **/app/code/core/Magento/Newsletter/controllers/SubscriberController.php** in function newAction()

```php
public function newAction()
    {
        if (!$this->_validateFormKey()) {
            $this->_redirectReferer();
            return;
        }

        if ($this->getRequest()->isPost() && $this->getRequest()->getPost('email')) {
            $session            = Mage::getSingleton('core/session');
            $customerSession    = Mage::getSingleton('customer/session');
            $email              = (string) $this->getRequest()->getPost('email');

            try {
                if (!Zend_Validate::is($email, 'EmailAddress')) {
                    Mage::throwException($this->__('Please enter a valid email address.'));
                }

                if (Mage::getStoreConfig(Mage_Newsletter_Model_Subscriber::XML_PATH_ALLOW_GUEST_SUBSCRIBE_FLAG) != 1 &&
                    !$customerSession->isLoggedIn()) {
                    Mage::throwException($this->__('Sorry, but administrator denied subscription for guests. Please <a href="%s">register</a>.', Mage::helper('customer')->getRegisterUrl()));
                }

                $ownerId = Mage::getModel('customer/customer')
                        ->setWebsiteId(Mage::app()->getStore()->getWebsiteId())
                        ->loadByEmail($email)
                        ->getId();
                if ($ownerId !== null && $ownerId != $customerSession->getId()) {
                    Mage::throwException($this->__('This email address is already assigned to another user.'));
                }

                $status = Mage::getModel('newsletter/subscriber')->subscribe($email);
                if ($status == Mage_Newsletter_Model_Subscriber::STATUS_NOT_ACTIVE) {
                    $session->addSuccess($this->__('Confirmation request has been sent.'));
                } else {
                    $session->addSuccess($this->__('Thank you for your subscription.'));
                }
            } catch (Mage_Core_Exception $e) {
                $session->addException($e, $this->__('There was a problem with the subscription: %s', $e->getMessage()));
            } catch (Exception $e) {
                $session->addException($e, $this->__('There was a problem with the subscription.'));
            }
        }
        $this->_redirectReferer();
    }
```

As can be seen, there is no provision in which the address is already subscribed. No error message is displayed to the visitor informing him that the address filled in is already subscribed and stoping the sending of the confirmation message.

**Comment by ADDISON74** [comment]
The code in function **newAction()**around line 65 should fix this issue:

```php
// check if guest email is subscribed already
$existingSubscriber = Mage::getModel('newsletter/subscriber')->loadByEmail($email)->getId();
if (!empty($existingSubscriber) && !$customerSession->isLoggedIn()):
    Mage::throwException($this->__('This email address is already subscribed.'));
endif;
```

**Comment by fballiano** [comment]
exactly, to me that part of code should already check if a user is already subscribed

**Comment by ADDISON74** [comment]
@fballiano - I leave to your discretion the name of the $existingSubscriber variable and what else we can take from the above code in a new PR.

**Comment by ADDISON74** [comment]
... maybe we need new phrases for adding to locale files.
